app/src/main/java/org/covidwatch/android/CovidWatchTcnManager.kt

```diff
@@ -0,0 +1,139 @@
+package org.covidwatch.android
+
+import android.app.Notification
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.app.PendingIntent
+import android.content.Context
+import android.content.Intent
+import android.os.Build
+import androidx.core.app.NotificationCompat
+import androidx.core.app.NotificationManagerCompat
+import androidx.core.content.ContextCompat
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import org.covidwatch.android.data.TemporaryContactNumber
+import org.covidwatch.android.data.TemporaryContactNumberDAO
+import org.covidwatch.android.data.signedreport.SignedReport
+import org.covidwatch.android.data.signedreport.SignedReportDAO
+import org.covidwatch.android.presentation.MainActivity
+import org.tcncoalition.tcnclient.TcnKeys
+import org.tcncoalition.tcnclient.TcnManager
+import java.util.*
+
+class CovidWatchTcnManager(
+    private val context: Context,
+    private val tcnKeys: TcnKeys,
+    private val tcnDao: TemporaryContactNumberDAO,
+    private val signedReportDAO: SignedReportDAO
+) : TcnManager(context) {
+    private val advertisedTcns = mutableListOf<ByteArray>()
+
+    override fun foregroundNotification(): NotificationCompat.Builder {
+        createNotificationChannelIfNeeded()
+
+        val notificationIntent = Intent(context, MainActivity::class.java)
+        val pendingIntent = PendingIntent.getActivity(
+            context,
+            0,
+            notificationIntent,
+            PendingIntent.FLAG_UPDATE_CURRENT
+        )
+
+        return NotificationCompat.Builder(
+            context,
+            CHANNEL_ID
+        )
+            .setContentTitle(context.getString(R.string.foreground_notification_title))
+            .setSmallIcon(R.drawable.ic_launcher_foreground)
+            .setContentIntent(pendingIntent)
+            .setPriority(NotificationManagerCompat.IMPORTANCE_HIGH)
+            .setCategory(Notification.CATEGORY_SERVICE)
+    }
+
+    override fun generateTcn(): ByteArray {
+        val tcn = tcnKeys.generateTcn()
+        advertisedTcns.add(tcn)
+        if (advertisedTcns.size > 65535) {
```

Comment:
magic const?

Reply:
Yeah...it's been there since the begging of the project I guess. We have it in the `develop` branch as well. 
For this PR i mostly copied the implementation from the `minimal` in order to minimize possible copypasta with edits issues.